src/pages/index.tsx

```diff
@@ -4,19 +4,108 @@ import { Layout } from "../components/Layout";
 import { useRouter } from "next/router";
 import dynamic from "next/dynamic";
 import { Posts } from "../components/Post";
+import { api } from "../utils/api";
+import { useEffect, useState } from "react";
+import { BlogPostCategory } from "../types/blogPost";
+import SearchPost from "../components/SearchPost";
+import { usePagination } from "../hooks/usePagination";
+import { useDebounce } from "../hooks/useDebounce";
+import { Pagination } from "../components/Pagination";
 
 const HomePage: NextPage = () => {
+  const [paginationPage, setPaginationPage] = useState(0);
+  const [searchActive, SetSearchActive] = useState(false);
+  const [searchCategory, setSearchCategory] = useState<BlogPostCategory[]>([]);
+  const [searchWord, setSearchWord] = useState<string>("");
+  const debounce = useDebounce(searchWord, 500);
   const router = useRouter();
+
+  useEffect(() => {
+    const params = new URLSearchParams(window.location.search);
+
+    const searchParam = params.get("search");
+    if (searchParam) {
+      setSearchWord(searchParam);
+    }
+
+    const categoryParam = params.get("category");
+    if (categoryParam !== null) {
+      const validCategories = Object.values(BlogPostCategory);
+      if (validCategories.includes(categoryParam as BlogPostCategory)) {
+        const categoryArray: BlogPostCategory[] = [
+          categoryParam as BlogPostCategory,
+        ];
+        setSearchCategory(categoryArray);
+      }
+    }
+  }, []);
+
+  const POSTS_PER_PAGE = 8;
+
+  const { data, isLoading, error } = api.blogPost.getBlogPosts.useQuery({
+    page: paginationPage,
+    perPage: POSTS_PER_PAGE,
+    categories: searchCategory,
+    search: debounce,
+  });
+  const handlePaginationPageChange = (paginationPage: number) => {
+    setPaginationPage(paginationPage);
+  };
+  const paginationPosts = usePagination(paginationPage, data?.total);
+  const handleSearchCategory = (category: BlogPostCategory[]) => {
+    setSearchCategory(category);
+    history.pushState(
+      null,
+      "search",
+      `?search=${searchWord}&category=${searchCategory}`
```

Comment:
What if you had 10 different params?

You can use this method instead
```
 await router.push({ query: blogPostFilters }, undefined, { shallow: true });
```

I haven't spend to much time on it, so this code may be wrong. Make sure it does what it is supposed to do (by reading next.js router documentation)